src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs

```diff
@@ -338,17 +316,75 @@ public void InitializeService(ServiceHost serviceHost)
 
         #endregion
 
+        #region Inter-Service API Handlers
+
+        /// <summary>
+        /// Query execution meant to be called from another service. Utilizes callbacks to allow
+        /// custom actions to be taken upon creation of query and failure to create query.
+        /// </summary>
+        /// <param name="executeParams">Params for creating the new query</param>
+        /// <param name="eventSender">Object that can send events for query execution progress</param>
+        /// <param name="queryCreatedAction">
+        /// Action to perform when query has been successfully created, right before execution of 
+        /// the query
+        /// </param>
+        /// <param name="failureAction">Action to perform if query was not successfully created</param>
+        public async Task InterServiceExecuteQuery(ExecuteRequestParamsBase executeParams, IEventSender eventSender,
+            Func<Task> queryCreatedAction, Func<string, Task> failureAction)
+        {
+            // Get a new active query
+            Query newQuery = await CreateAndActivateNewQuery(executeParams, queryCreatedAction, failureAction);
+
+            // Execute the query asynchronously
+            ExecuteAndCompleteQuery(executeParams.OwnerUri, eventSender, newQuery);
+        }
+
+        /// <summary>
+        /// Query disposal meant to be called from another service. Utilizes callbacks to allow
+        /// custom actions to be performed on success or failure.
+        /// </summary>
+        /// <param name="ownerUri">The identifier of the query to be disposed</param>
+        /// <param name="successAction">Action to perform on success</param>
+        /// <param name="failureAction">Action to perform on failure</param>
+        /// <returns></returns>
+        public async Task InterServiceDisposeQuery(string ownerUri, Func<Task> successAction,
+            Func<string, Task> failureAction)
+        {
+            try
+            {
+                // Attempt to remove the query for the owner uri
+                Query result;
+                if (!ActiveQueries.TryRemove(ownerUri, out result))
+                {
+                    await failureAction(SR.QueryServiceRequestsNoQuery);
+                    return;
+                }
+
+                // Cleanup the query
+                result.Dispose();
+
+                // Success
+                await successAction();
```

Comment:
To prevent an `await null` we should either check for null here or in `InterServiceExecuteQuery`. Likewise for `failureAction`.